### PR TITLE
Test various upgrade scenarios

### PR DESCRIPTION
**Pull request by sbodagala**
# Description

Write tests that cover the following scenarios:

- Test status json in the context of version incompatible upgrades.

- Test that restarts multiple processes (a storage process and multiple stateless processes) during the stage phase

- Test that tests cluster generation number during upgrade

## Type of change

- Other (adds an upgrade test; no code changes)

## Discussion

*Are there any design details that you would like to discuss further?*
No

## Testing

Ran the test manually.

## Documentation

*Did you update relevant documentation within this repository?*
N/A

*If this change is adding new functionality, do we need to describe it in our user manual?*
N/A

*If this change is adding or removing subreconcilers, have we updated the core technical design doc to reflect that?*
N/A

*If this change is adding new safety checks or new potential failure modes, have we documented and how to debug potential issues?*
N/A

## Follow-up

*Are there any follow-up issues that we should pursue in the future?*
No

*Does this introduce new defaults that we should re-evaluate in the future?*
No

### CODE REVIEW

**Comment by sbodagala** [comment]
Ran the test manually, and it failed with this error: "invariant InvariantClusterStatusAvailableWithThreshold failed".

It appears to me the upgrade completed successfully. Here's the output of "kubectl-fdb analyze" over the cluster after the test failed with the above error:

> # kubectl-fdb analyze fdb-cluster-cz8blukk -n sre-s3xfe49r
> Checking cluster: sre-s3xfe49r/fdb-cluster-cz8blukk
> ✔ Cluster is available
> ✔ Cluster is fully replicated
> ✔ Cluster is reconciled
> ✔ ProcessGroups are all in ready condition
> ✔ Pods are all running and available
> Checking cluster: sre-s3xfe49r/fdb-cluster-cz8blukk with auto-fix: false

File: e2e/test_operator_upgrades/operator_upgrades_test.go [hunk 1]
```diff
@@ -376,6 +379,22 @@ var _ = Describe("Operator Upgrades", Label("e2e"), func() {
 			)
 			Expect(err).NotTo(HaveOccurred())
 
+			// Wait for the server process to restart.
+			time.Sleep(40 * time.Second)
```
Comment (johscheuer) [inline e2e/test_operator_upgrades/operator_upgrades_test.go:383]: Are those changes intended to be in this PR? They seem to be unrelated to the initial changes "test upgrading a cluster when no storage processes are restarted"?

File: e2e/test_operator_upgrades/operator_upgrades_test.go [hunk 2]
```diff
@@ -376,6 +379,22 @@ var _ = Describe("Operator Upgrades", Label("e2e"), func() {
...
+
+			// Check if the restarted process is showing up in IncompatibleConnections list in status output.
+			status := fdbCluster.GetStatus()
+			log.Println("IncompatibleProcesses:", status.Cluster.IncompatibleConnections)
+
+			// NOTE: The restarted process doesn't show up in IncompatibleConnections list consistently, hence
```
Comment (johscheuer) [inline e2e/test_operator_upgrades/operator_upgrades_test.go:389]: Could you open an issue for that? We should investigate this issue on FDB side and make sure we understand why the connection is not appearing in the incompatible connection list consistently. And does `The restarted process doesn't show up in IncompatibleConnections list consistently` mean that 1.) It's not appearing at all in the list? 2.) It's in a state where the address is appearing and disappearing? If I remember correct the detection time was something around 100 seconds, but I would need to review the FDB code. again.

File: e2e/test_operator_upgrades/operator_upgrades_test.go [hunk 3]
```diff
@@ -386,6 +405,72 @@ var _ = Describe("Operator Upgrades", Label("e2e"), func() {
 		fixtures.GenerateUpgradeTableEntries(testOptions),
 	)
 
+	DescribeTable(
+		"upgrading a cluster where a storage and multiple stateless processes get restarted during the staging phase",
```
Comment (johscheuer) [inline e2e/test_operator_upgrades/operator_upgrades_test.go:409]: This seems to be a completely unrelated test case to the PR name? Can we open a new PR for this test?

File: e2e/test_operator_upgrades/operator_upgrades_test.go [hunk 4]
```diff
@@ -481,6 +568,12 @@ var _ = Describe("Operator Upgrades", Label("e2e"), func() {
 
 			// The cluster should still be able to upgrade.
 			Expect(fdbCluster.UpgradeCluster(targetVersion, true)).NotTo(HaveOccurred())
+
+			// NOTE: The coordinator process selected above is getting restarted - why the buggify option
```
Comment (johscheuer) [inline e2e/test_operator_upgrades/operator_upgrades_test.go:572]: The coordinator process is not being restarted during the `kill` command (bounce phase) but in the later phase where all Pods are recreated. Assuming the coordinator is coming up in the incompatible connection list  this sub reconciler: https://github.com/FoundationDB/fdb-kubernetes-operator/blob/main/controllers/remove_incompatible_processes.go will recreate the Pod.

File: e2e/test_operator_upgrades/operator_upgrades_test.go [hunk 5]
```diff
@@ -874,4 +967,52 @@ var _ = Describe("Operator Upgrades", Label("e2e"), func() {
...
+
+			log.Println(
+				"Selected Pods:",
+				ignoreDuringRestart,
+				"to be skipped during the restart",
+			)
+			fdbCluster.SetIgnoreDuringRestart(ignoreDuringRestart)
+
+			// The cluster should still be able to upgrade.
+			Expect(fdbCluster.UpgradeCluster(targetVersion, true)).NotTo(HaveOccurred())
```
Comment (johscheuer) [inline e2e/test_operator_upgrades/operator_upgrades_test.go:1012]: Are you sure the upgrade will succeed if no storage server is upgraded? That would probably result in "dataloss" and that should block the operator from doing further actions.

Comment (sbodagala) [reply e2e/test_operator_upgrades/operator_upgrades_test.go:1012]: The output of "kubectl-fdb analyze" over the cluster after running the test: https://github.com/FoundationDB/fdb-kubernetes-operator/pull/1580#issuecomment-1501916726

The test did throw an error though (which I mentioned in https://github.com/FoundationDB/fdb-kubernetes-operator/pull/1580#issuecomment-1501916726).

I don't have any more insights into this that what is reported by "kubectl-fdb analyze".

Maybe we should have a theory as to what is supposed to happen in this scenario and verify whether the test output matches that theory or not (rather than simply running the test and observe the outcome, which we are trying to do here).

File: e2e/test_operator_upgrades/operator_upgrades_test.go (hunk 1 above)
Comment (sbodagala) [reply e2e/test_operator_upgrades/operator_upgrades_test.go:383]: Adjusted the scope of this PR to cover testing multiple upgrade scenarios.

File: e2e/test_operator_upgrades/operator_upgrades_test.go (hunk 2 above)
Comment (sbodagala) [reply e2e/test_operator_upgrades/operator_upgrades_test.go:389]: Let me try changing "time.Sleep()" to 100 seconds and see what happens.

File: e2e/test_operator_upgrades/operator_upgrades_test.go (hunk 3 above)
Comment (sbodagala) [reply e2e/test_operator_upgrades/operator_upgrades_test.go:409]: As I mentioned above, extended the scope of this PR to cover all these cases.

File: e2e/test_operator_upgrades/operator_upgrades_test.go (hunk 4 above)
Comment (sbodagala) [reply e2e/test_operator_upgrades/operator_upgrades_test.go:572]: Ok, I will remove this comment then. Thanks!

File: e2e/test_operator_upgrades/operator_upgrades_test.go (hunk 5 above)
Comment (sbodagala) [reply e2e/test_operator_upgrades/operator_upgrades_test.go:1012]: I ran this test again. (The behavior is similar to what I reported here: https://github.com/FoundationDB/fdb-kubernetes-operator/pull/1580#issuecomment-1501916726).

The test failed with this error:

> 
> <*errors.errorString | 0xc002b67680>: {
>           s: "invariant InvariantClusterStatusAvailableWithThreshold failed",
>       }
>       invariant InvariantClusterStatusAvailableWithThreshold failed
>   occurred
> 

Output of "kubectl-fdb analyze":

```
# kubectl-fdb analyze fdb-cluster-in36toff -n  sre-nawokdnf
Checking cluster: sre-nawokdnf/fdb-cluster-in36toff
✔ Cluster is available
✔ Cluster is fully replicated
✔ Cluster is reconciled
✔ ProcessGroups are all in ready condition
✔ Pods are all running and available
Checking cluster: sre-nawokdnf/fdb-cluster-in36toff with auto-fix: false
```

The output of "kubectl -n  sre-nawokdnf get fdb,po" shows that the storage processes are running at 7.1.27 version (I checked the version of fdbserver binary in one of the storage pods).

File: e2e/test_operator_upgrades/operator_upgrades_test.go [hunk 6]
```diff
@@ -874,4 +959,52 @@ var _ = Describe("Operator Upgrades", Label("e2e"), func() {
 		EntryDescription("Upgrade from %[1]s to %[2]s and one process is missing the new binary"),
 		fixtures.GenerateUpgradeTableEntries(testOptions),
 	)
+
+	DescribeTable(
+		"upgrading a cluster when no storage processes are restarted",
+		func(beforeVersion string, targetVersion string) {
+			// We set the before version here to overwrite the before version from the specific flag
```
Comment (johscheuer) [inline e2e/test_operator_upgrades/operator_upgrades_test.go:966]: Comment can be removed

File: e2e/test_operator_upgrades/operator_upgrades_test.go (hunk 5 above)
Comment (johscheuer) [reply e2e/test_operator_upgrades/operator_upgrades_test.go:1004]: The analyze command is mostly reading the information from the FoundationDB cluster object and checking the Pods. We should check the operator logs to get a better understanding of the behaviour.

**Review by johscheuer — CHANGES_REQUESTED** [review]
Could we split up the tests into separate PRs? Otherwise one single test will block the PR from being merged.

File: e2e/test_operator_ha_upgrades/operator_ha_upgrade_test.go [hunk 7]
```diff
@@ -381,4 +381,38 @@ var _ = Describe("Operator HA Upgrades", Label("e2e"), func() {
...
+		func(beforeVersion string, targetVersion string) {
+			clusterSetup(beforeVersion, false)
+
+			initialGeneration := 0
+			for _, singleCluster := range fdbCluster.GetAllClusters() {
+				status := singleCluster.GetStatus()
+				if status.Cluster.Generation > initialGeneration {
+					initialGeneration = status.Cluster.Generation
+				}
+			}
```
Comment (johscheuer) [inline e2e/test_operator_ha_upgrades/operator_ha_upgrade_test.go:396]:
```suggestion
			status := fdbCluster.GetPrimary().GetStatus()
			initialGeneration = status.Cluster.Generation
```

File: e2e/test_operator_ha_upgrades/operator_ha_upgrade_test.go [hunk 8]
```diff
@@ -381,4 +381,38 @@ var _ = Describe("Operator HA Upgrades", Label("e2e"), func() {
...
+
+			// Start the upgrade, but do not wait for reconciliation to complete.
+			Expect(fdbCluster.UpgradeCluster(targetVersion, false)).NotTo(HaveOccurred())
+
+			Eventually(func() bool {
+				for _, singleCluster := range fdbCluster.GetAllClusters() {
```
Comment (johscheuer) [inline e2e/test_operator_ha_upgrades/operator_ha_upgrade_test.go:402]: Should use `fdbCluster.GetPrimary().GetStatus()`

File: e2e/test_operator_ha_upgrades/operator_ha_upgrade_test.go [hunk 9]
```diff
@@ -381,4 +381,38 @@ var _ = Describe("Operator HA Upgrades", Label("e2e"), func() {
...
+			initialGeneration := 0
+			for _, singleCluster := range fdbCluster.GetAllClusters() {
+				status := singleCluster.GetStatus()
+				if status.Cluster.Generation > initialGeneration {
+					initialGeneration = status.Cluster.Generation
+				}
+			}
+
+			// Start the upgrade, but do not wait for reconciliation to complete.
+			Expect(fdbCluster.UpgradeCluster(targetVersion, false)).NotTo(HaveOccurred())
```
Comment (johscheuer) [inline e2e/test_operator_ha_upgrades/operator_ha_upgrade_test.go:399]:
```suggestion
			Expect(fdbCluster.UpgradeCluster(targetVersion, true)).NotTo(HaveOccurred())
```

File: e2e/test_operator_ha_upgrades/operator_ha_upgrade_test.go [hunk 10]
```diff
@@ -381,4 +381,38 @@ var _ = Describe("Operator HA Upgrades", Label("e2e"), func() {
 		EntryDescription("Upgrade from %[1]s to %[2]s with network link that drops some packets"),
 		fixtures.GenerateUpgradeTableEntries(testOptions),
 	)
+
+	DescribeTable(
+		"Test ha cluster generation number during upgrade",
```
Comment (johscheuer) [inline e2e/test_operator_ha_upgrades/operator_ha_upgrade_test.go:386]: We will merge the changes/checks from here into `upgrading a cluster with operator pod chaos and without foundationdb pod chaos`

File: e2e/test_operator_upgrades/operator_upgrades_test.go (hunk 4 above)
Comment (sbodagala) [reply e2e/test_operator_upgrades/operator_upgrades_test.go]: Done.

File: e2e/test_operator_upgrades/operator_upgrades_test.go (hunk 5 above)
Comment (sbodagala) [reply e2e/test_operator_upgrades/operator_upgrades_test.go]: Removed this test from this PR. Will make another PR for that test.

File: e2e/test_operator_upgrades/operator_upgrades_test.go [hunk 11]
```diff
@@ -386,6 +400,72 @@ var _ = Describe("Operator Upgrades", Label("e2e"), func() {
 		fixtures.GenerateUpgradeTableEntries(testOptions),
 	)
 
+	DescribeTable(
+		"upgrading a cluster where a storage and multiple stateless processes get restarted during the staging phase",
+		func(beforeVersion string, targetVersion string) {
+			// We set the before version here to overwrite the before version from the specific flag
```
[the remark this answers is not in the document]
Comment (sbodagala) [reply e2e/test_operator_upgrades/operator_upgrades_test.go]: Done.

File: e2e/test_operator_upgrades/operator_upgrades_test.go [hunk 12]
```diff
@@ -481,6 +563,9 @@ var _ = Describe("Operator Upgrades", Label("e2e"), func() {
 
 			// The cluster should still be able to upgrade.
 			Expect(fdbCluster.UpgradeCluster(targetVersion, true)).NotTo(HaveOccurred())
+
+			status := fdbCluster.GetStatus()
+			Expect(len(status.Cluster.IncompatibleConnections)).To(Equal(0))
```
[the remark this answers is not in the document]
Comment (sbodagala) [reply e2e/test_operator_upgrades/operator_upgrades_test.go:569]: No, I was actually trying to test the scenario where a process is not getting restarted and verify that that process is showing up in incompatibleConnections list. But all processes, including the coordinator that was prevented from restart by using the buggify option, are getting restarted at this point.

File: e2e/test_operator_upgrades/operator_upgrades_test.go (hunk 6 above)
Comment (sbodagala) [reply e2e/test_operator_upgrades/operator_upgrades_test.go]: This test has been removed.

File: e2e/test_operator_ha_upgrades/operator_ha_upgrade_test.go (hunk 10 above)
Comment (sbodagala) [reply e2e/test_operator_ha_upgrades/operator_ha_upgrade_test.go]: Done.

**Comment by sbodagala** [comment]
Uploaded the latest version, please take a look. Thanks!

**Comment by sbodagala** [comment]
> ### Result of fdb-kubernetes-operator-pr on Linux CentOS 7
> * Commit ID: [e379366](https://github.com/FoundationDB/fdb-kubernetes-operator/commit/e3793665f5339fc5b90406edac27c92a7eeef9b2)
> * Duration 2:50:59
> * Result: ❌ **FAILED**
> * Error: `Error while executing command: if $fail_test; then exit 1; fi. Reason: exit status 1`
> * [Build Log](https://d1e0l78xbkh3xa.cloudfront.net/fdb-kubernetes-operator/fdb-kubernetes-operator-pr/02cff1e1-e1b8-431f-9a96-2f52f2956ac1.gz) terminal output (available for 30 days)
> * [Build Workspace](https://d1e0l78xbkh3xa.cloudfront.net/fdb-kubernetes-operator-pr/02cff1e1-e1b8-431f-9a96-2f52f2956ac1/workspace.zip) zip file of the working directory (available for 30 days)

Reports the following failures:

operator_ha_upgrade_test.go: 

Test "upgrading a cluster with operator pod chaos and without foundationdb pod chaos"" failed because "Cluster.Generation" after upgrade is 37, instead of 19.

NOTE: I modified "Upgrading a multi-DC cluster without chaos" test to check "Cluster.Generation" after upgrade - again, "Cluster.Generation" is 34, instead of 19.

operator_upgrades_test.go: 

Test "upgrading a cluster where a storage and multiple stateless processes get restarted during the staging phase Upgrade" failed on this error:

> <*errors.errorString | 0xc0006c1220>: {
>           s: "invariant InvariantClusterStatusAvailableWithThreshold failed",
>       }
>       invariant InvariantClusterStatusAvailableWithThreshold failed

**Comment by sbodagala** [comment]
More on the failure in operator_ha_upgrade_test.go: I don't see any helpful information in the test output, but I do see recoveries (on Splunk) seconds apart (like these: 2023-05-01T16:01:30Z and 2023-05-01T16:01:32Z; these are the timestamps of Type "MasterRecoveryState" with "Status:  reading_coordinated_state"). I think this is the result of server processes getting bounced at different, but relatively closer, timestamps.

**Comment by sbodagala** [comment]
> Test "upgrading a cluster where a storage and multiple stateless processes get restarted during the staging phase Upgrade" failed on this error:
> 
> <*errors.errorString | 0xc0006c1220>: {
> s: "invariant InvariantClusterStatusAvailableWithThreshold failed",
> }
> invariant InvariantClusterStatusAvailableWithThreshold failed

Ran this test locally multiple times, they all succeeded. So the failure reported by CI might not be related to this specific test.

**Comment by johscheuer** [comment]
Seems like the last test run hit some issues. I try another run.

File: e2e/test_operator_ha_upgrades/operator_ha_upgrade_test.go [hunk 13]
```diff
@@ -168,7 +168,14 @@ var _ = Describe("Operator HA Upgrades", Label("e2e"), func() {
 		"upgrading a cluster with operator pod chaos and without foundationdb pod chaos",
 		func(beforeVersion string, targetVersion string) {
 			clusterSetup(beforeVersion, true /* = enableOperatorPodChaos */)
+			initialGeneration := fdbCluster.GetPrimary().GetStatus().Cluster.Generation
 			upgradeAndVerify(fdbCluster, targetVersion)
+			// Verify that the cluster generation number didn't increase by more
+			// than 40 (in an ideal case the number of recoveries that should happen
+			// during an upgrade is 9, but in reality that number could be higher
+			// because different server processes may get bounced at different times).
+			Expect(fdbCluster.GetPrimary().GetStatus().Cluster.Generation).To(BeNumerically("<=", initialGeneration+40))
```
Comment (johscheuer) [inline e2e/test_operator_ha_upgrades/operator_ha_upgrade_test.go:177]: Could you create an issue for this? We should follow up on this, to understand why we see a higher number of generations than expected.

**Review by johscheuer — COMMENTED** [review]
One the e2e test pipeline passes we can merge this PR 👍

**Comment by johscheuer** [comment]
> ### Result of fdb-kubernetes-operator-pr on Linux CentOS 7
> * Commit ID: [2b65a07](https://github.com/FoundationDB/fdb-kubernetes-operator/commit/2b65a07f343d7e1d90ebbfb037d15a03608659e6)
> * Duration 3:56:49
> * Result: ❌ **FAILED**
> * Error: `Error while executing command: if $fail_test; then exit 1; fi. Reason: exit status 1`
> * [Build Log](https://d1e0l78xbkh3xa.cloudfront.net/fdb-kubernetes-operator/fdb-kubernetes-operator-pr/77044905-b2de-4cc7-9f2a-5b4c69b81964.gz) terminal output (available for 30 days)
> * [Build Workspace](https://d1e0l78xbkh3xa.cloudfront.net/fdb-kubernetes-operator-pr/77044905-b2de-4cc7-9f2a-5b4c69b81964/workspace.zip) zip file of the working directory (available for 30 days)

```
Summarizing 1 Failure:
  [FAIL] Operator Upgrades [AfterEach] upgrading a cluster where a storage and multiple stateless processes get restarted during the staging phase Upgrade from 6.3.25 to 7.1.27 with a storage and multiple stateless processes restarted during the staging phase [e2e]
  /codebuild/output/src828282037/src/github.com/FoundationDB/fdb-kubernetes-operator/e2e/fixtures/fixtures.go:62
```

File: e2e/test_operator_upgrades/operator_upgrades_test.go [hunk 14]
```diff
@@ -386,6 +403,70 @@ var _ = Describe("Operator Upgrades", Label("e2e"), func() {
...
+			Expect(fdbCluster.UpgradeCluster(targetVersion, false)).NotTo(HaveOccurred())
+
+			// Wait until all process groups are in the staging phase and the new binaries are available.
+			Eventually(func() bool {
+				return fdbCluster.AllProcessGroupsHaveCondition(fdbv1beta2.IncorrectCommandLine)
+			}).WithTimeout(10 * time.Minute).WithPolling(2 * time.Second).Should(BeTrue())
+
+			// Restart the processes selected above to pickup the new configuration and run with the
+			// newer version.
+			for _, selectedPod := range selectedPods {
```
Comment (johscheuer) [inline e2e/test_operator_upgrades/operator_upgrades_test.go:450]: What is the intention of this test case? Right now I don't see any value in adding this test case:

1.) We set those processes to not be restarted
2.) We start the upgrade
3.) We check that the processes are getting the `IncorrectCommandLine` condition 
4.) We directly restart the affected processes

Shouldn't we do some kind of validation between 3. and 4.? Otherwise this is the same as doing a "normal" upgrade with additional steps.

If we have a theory/hypothesis we want to prove we should document that and make the according assertions. Otherwise I think we should remove this test case. What do you think?

**Review by johscheuer — CHANGES_REQUESTED** [review]

File: e2e/test_operator_upgrades/operator_upgrades_test.go (hunk 14 above)
Comment (sbodagala) [reply e2e/test_operator_upgrades/operator_upgrades_test.go:450]: I was merely following what is specified in the radar, don't have a deeper understanding of what this test is supposed to accomplish. Let's delete this test then.

File: e2e/test_operator_ha_upgrades/operator_ha_upgrade_test.go (hunk 13 above)
Comment (sbodagala) [reply e2e/test_operator_ha_upgrades/operator_ha_upgrade_test.go:177]: Ok.

**Review by johscheuer — APPROVED** [review]

File: e2e/test_operator_ha_upgrades/operator_ha_upgrade_test.go (hunk 13 above)
Comment (sbodagala) [reply e2e/test_operator_ha_upgrades/operator_ha_upgrade_test.go:177]: Created issue: https://github.com/FoundationDB/fdb-kubernetes-operator/issues/1607
